### DISCUSSION TITLE
Resolving `ordinal_`  imports from mfc140.dll

### DISCUSSION THREAD

**Discussion by randomstr1ng** [Vector35/binaryninja-api#6091] [Analysis]
Currently I working on a Windows Application which has a single main executeable, a bunch of it's own dll's and in addition it is using a bunch of system dll's like `kernel32.dll`, `user32.dll`, and so on.

While analysing the program, I have a lot of calls for `ordinal_mfc140_XXXX` like below
![image](https://github.com/user-attachments/assets/6a8afaa2-58a0-446b-86c3-4938683f278a)
In addition, I can see that there is no Type Library or anything reagards the `mfc140.dll` loaded.
![image](https://github.com/user-attachments/assets/06393f53-bf6b-4197-ab98-2c372988082b)

I have loaded the same dll into Ghidra (on the same Windows machine) to see how it handles the calls. Ghidra is able to load the exports and resolve those.
![image](https://github.com/user-attachments/assets/a78c3a6c-5838-4338-87c5-24aeac2ad5b6)

Is there a elegant way how I can resolve this/get the same behavior like Ghidra has with Binary Ninja?
Can I export the calls in Ghidra and import them into Binary Ninja?

Thanks!

**Comment by plafosse** [comment]
Yes its pretty easy todo but you need to write some code.

```
from binaryninja import *

tl = TypeLibrary.new(Architecture['x86'], 'mfc140')
my_ordinals = [
    ("name", 0x1000),
]

ordinal_version = "ordinals_v1"
tl.store_metadata("ordinals", ordinal_version)
tl.store_metadata(ordinal_version, my_ordinals)
tl.write_to_file("mfc140.bntl")
```
Then you need to install this bntl in your:  `<Binary Ninja Home directory>/typlib/x86/`

and I *think* that will work. We really do need to generate type libraries for all the MFC stuff

**Comment by plafosse** [comment]
As we're in the middle of a release process I don't have the time to add these to the product and test them but you these may be helpful to you:

[MFCOrdinalTypeLibs.zip](https://github.com/user-attachments/files/17636806/MFCOrdinalTypeLibs.zip)

**Comment by randomstr1ng** [comment]
Hi @plafosse,
thank you for your response and for all the .bntl files!

I generated my own .bntl as you had described above and placed it in the folder but for what ever reason it does not show up.
The same happens with the files you send me in the archive.

But, I did a debug log and I can see, that the mfc140.dll.bntl file is loaded from my location and it also uses it in the analysis:
![image](https://github.com/user-attachments/assets/e41463bc-c471-40d1-b209-5ebe87b30ffb)
![image](https://github.com/user-attachments/assets/ec7c0e2f-f2cd-4b38-be96-b1220070050b)

If I have a look at the loaded types in the UI, the `mfc140.dll` is still missing, also the Ordinal call are there and have not been resolved.
![image](https://github.com/user-attachments/assets/2a0c6f6e-59d7-4e3f-b7c7-d2fbd0d9adba)

Is there a platform mismatch or something? Appreciate any hint as I know you are busy with the release process.
Let me know if the debug log would help. (~100MB)

**Comment by plafosse** [comment]
Any chance you can share some binaries?

**Reply by randomstr1ng** [reply]
@plafosse sure - try this one: https://1drv.ms/u/s!AsS7Wq-A6cTAvDCcBYE6Zdm6Dvxf?e=MfqphT
(seems like the last upload had been already deleted.

thank you in advance!

**Comment by randomstr1ng** [comment]
@plafosse did you had a chance to have a look?

**Comment by pjcreath** [comment]
@plafosse Bump. I ran into this with 5.2.8722-Stable. Adding the following seemed to fix its absence from the "Add Type Library" menu, but after adding it manually, it shows up with 0 entries:

```
tl.add_platform(Platform["windows-x86_64"])
```

I notice other bntl files have named_objects and named_types, not just ordinals in the metadata?
